README.md
# avro-rs

[![GitHub version](https://badge.fury.io/gh/flavray%2Favro-rs.svg)](https://badge.fury.io/gh/flavray%2Favro-rs)
[![Build Status](https://travis-ci.org/flavray/avro-rs.svg?branch=master)](https://travis-ci.org/flavray/avro-rs)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/flavray/avro-rs/blob/master/LICENSE)

[Documentation](https://docs.rs/avro-rs)

A library for working with [Apache Avro](https://avro.apache.org/) in Rust.

Please check our [documentation](https://docs.rs/avro-rs) for examples, tutorials and API reference.

## Example

Add to your `Cargo.toml`:

```toml
[dependencies]
avro-rs = "0.1"
```

Then try to write and read in Avro format like below:

```rust
extern crate avro;

#[macro_use]
extern crate serde_derive;
extern crate failure;

use avro::{Codec, Reader, Schema, Writer, from_value, types::Record};
use failure::Error;

#[derive(Debug, Deserialize, Serialize)]
struct Test {
 a: i64,
 b: String,
}

fn main() -> Result<(), Error> {
 let raw_schema = r#"
 {
 "type": "record",
 "name": "test",
 "fields": [
 {"name": "a", "type": "long", "default": 42},
 {"name": "b", "type": "string"}
 ]
 }
 "#;

 let schema = Schema::parse_str(raw_schema)?;

 println!("{:?}", schema);

 let mut writer = Writer::with_codec(&schema, Vec::new(), Codec::Deflate);

 let mut record = Record::new(writer.schema()).unwrap();
 record.put("a", 27i64);
 record.put("b", "foo");

 writer.append(record)?;

 let test = Test {
 a: 27,
 b: "foo".to_owned(),
 };

 writer.append_ser(test)?;

 writer.flush()?;

 let input = writer.into_inner();
 let reader = Reader::with_schema(&schema, &input[..])?;

 for record in reader {
 println!("{:?}", from_value::<Test>(&record?));
 }
 Ok(())
}
```

## License
This project is licensed under [MIT License](https://github.com/flavray/avro-rs/blob/master/LICENSE).
Please note that this is not an official project maintained by [Apache Avro](https://avro.apache.org/).

## Contributing
Everyone is encouraged to contribute! You can contribute by forking the GitHub repo and making a pull request or opening an issue.
All contributions will be licensed under [MIT License](https://github.com/flavray/avro-rs/blob/master/LICENSE).


CHANGES.txt
Avro Change Log

Trunk (not yet released)

 INCOMPATIBLE CHANGES

 AVRO-2003: Report specific location of schema incompatibilities (teabot via nkollar)

 AVRO-2035: Java: validate default values when parsing schemas. (cutting)

 AVRO-2043: Move to java8

 AVRO-2061: Improve Invalid File Format Error Message (Beluga Behr via gabor)

 AVRO-1933: Add more specific error details to SchemaCompatibility class
 (Anders Sundelin via gabor)

 AVRO-2072: ResolvingGrammarGenerator doesn't implement schema resolution correctly for unions
 (Nandor Kollar via gabor)

 AVRO-2059: Remove support of Hadoop 1 (gabor)

 NEW FEATURES

 AVRO-1704: Java: Add support for single-message encoding. (blue)

 AVRO-1704: Spec: Add single-message encoding format. (Niels Basjes via blue)

 AVRO-1884: Java: Add method to set the compiler output suffix.
 (shijinkui via blue)

 AVRO-1932: Java: Allow setting the SchemaStore on generated classes.
 (Niels Basjes)

 AVRO-2132: Allow dots in IDL property names. (kdrakon via cutting)

 OPTIMIZATIONS

 IMPROVEMENTS

 AVRO-1889: Update maven-shade-plugin to enable building in Java 8 on Mac.
 (Sachin Goyal via blue)

 AVRO-1695. Java: Fix GenericData#deepCopy() to support logical types.
 (cutting)

 AVRO-1719: C++: Fix tests with Boost >= 1.59. (Romain Geissler via blue)

 AVRO-1879: Make conversions field static. (Michael Wong via blue)

 AVRO-1919: Formalize the presence of build.sh. (Suraj Acharya via cutting)

 AVRO-1918: Improve logical type documentation. (Gabor Szadovszky via cutting)

 AVRO-1907: Java: Add debug logging for invalid logical types.
 (Gabor Szadovszky via blue)

 AVRO-1897: Fix build issues due to VERSION.txt newline, avro-tools.
 (Suraj Acharya via blue)

 AVRO-1993: C++ Byte ordering macro does not work on FreeBSD (thiru)

 AVRO-1975: Upgrade java dependencies (gabor)

 AVRO-1960: Add log4j properties for avro-tools

 AVRO-1748. Add Snappy Compression to C++ DataFile (J. Langley via thiru)

 AVRO-1723. Add support for forward declarations in avro IDL (Zoltan Farcas via thiru)

 AVRO-1971: Allow alias annotation on fields (Johannes Schulte via gabor)

 AVRO-2021: uuid logical type is not documented (Nandor Kollar via gabor)

 AVRO-2056: DirectBinaryEncoder Creates Buffer For Each Call To writeDouble
 (Beluga Behr via gabor)

 AVRO-2054: Use StringBuilder instead of StringBuffer
 (Beluga Behr via gabor)

 AVRO-2050: Clear Array To Allow GC
 (Beluga Behr via gabor)

 AVRO-2049: Remove Superfluous Configuration From AvroSerializer
 (Beluga Behr via gabor)

 AVRO-2048: Avro Binary Decoding - Gracefully Handle Long Strings
 (Beluga Behr via gabor)

 AVRO-2053: Remove Reference To Deprecated Property mapred.output.compression.type
 (Beluga Behr via gabor)

 AVRO-2055: Remove Magic Value From org.apache.avro.hadoop.io.AvroSequenceFile
 (Beluga Behr via gabor)

 AVRO-1047: Remaining warnings in generated Java code exposed by -Xlint:all fixed.
 (Michael Werle via Niels Basjes)

 AVRO-2080: Fix Javadoc Warnings in Generated Records
 (Bridger Howell via Niels Basjes)

 AVRO-2114: Make missing value exceptions in nested structures easier to read.
 (Niels Basjes)

 AVRO-2115: Java: Permit @Union annotations on fields of classes.
 (Miguel Martinez-Espronceda via cutting).

 AVRO-2119: Run Apache Rat check on every java build (Niels Basjes)

 AVRO-2117: Overall cleanup of java code (Niels Basjes)

 AVRO-2109: Reset buffers in case of IOException (gabor)

 AVRO-2133: Log rat failures (gabor)

 AVRO-2113: Improve unexpected type error message (thiru)

 BUG FIXES

 AVRO-1741: Python3: Fix error when codec is not in the header.
 (Matthew Hayes via blue)

 AVRO-1869: Java: Fix Decimal conversion from ByteBuffer. (blue)

 AVRO-607: Java: Make SpecificData schema cache thread-safe.
 (Andrius Druzinis-Vitkus via blue)

 AVRO-1888: Java: Fix single-record encoding marker check. (blue)

 AVRO-1848: Ruby: Fix handling of falsey default values.
 (Brian McKelvey via blue)

 AVRO-1874: Py3: Use recommended module-level logging.
 (Torgeir Børresen via blue)

 AVRO-1883: Java: Fix incompatible schema detection nested in unions.
 (Yibing Shi via blue)

 AVRO-1908: Fix TestSpecificCompiler reference to private method.
 (blue)

 AVRO-1873: Ruby: Add CRC32 checksum to Snappy-compressed blocks.
 (blue)

 AVRO-1900: Java: Remove dev-tools to fix Java build. (blue)

 AVRO-1913: No warnings in the generated Java code. (Niels Basjes)

 AVRO-1914: Fix licencing errors reported by rat. (Niels Basjes)

 AVRO-1901: Record named "Exception" generated bad code.
 (Radai Rosenblatt via Niels Basjes)

 AVRO-1849: C++: printJson fails on record with no fields.
 (Simon Woodford vai tomwhite)

 AVRO-1853: C++: Compiler::toBin crashes on empty string.
 (Zoltan Ivanfi via tomwhite)

 AVRO-1923, AVRO-1944: Stop infinite recursion in GenericData.toString. (Niels Basjes)

 AVRO-1860: Java: Field#DefaultVal() returns Ints for Long fields.
 (Gabor Szadovszky via cutting)

 AVRO-1929: Java: SchemaCompatibility fails to recognize
 string<->bytes promotion. (Anders Sundelin via cutting)

 AVRO-1882: Java: Fix ConcurrentHashMap with non-string keys.
 (Sachin Goyal via blue)

 AVRO-1943: Java: Flaky test:
 TestNettyServerWithCompression.testConnectionsCount
 (Gabor Szadovszky via tomwhite)

 AVRO-1947: Java: Inconsistent maven profile activation
 (Gabor Szadovszky via tomwhite)

 AVRO-1626: C#: Fix Avro.pref build error. (Naruto Takahashi via blue)

 AVRO-1951: Python: Fix ipc_test.py with a valid DNS name. (blue)

 AVRO-1966: Java: Fix NPE When copying builder with nullable record. (Niels Basjes)

 AVRO-1967: Java: Fix NPE when calling getXyzBuilder on instance where the xyz is null
 (Niels Basjes)

 AVRO-1970: Java: Flaky test: TestInputBytes. (Gabor Szadovszky via tomwhite)

 AVRO-1881: Java: Avro (Java) Memory Leak when reusing JsonDecoder instance.
 (Nandor Kollar via tomwhite)

 AVRO-1954: Java: Schema.Field.defaultVal() generates: Unknown datum type
 (Nandor Kollar via tomwhite)


src/lib.rs
//! ```
//! extern crate avro;
//!
//! #[macro_use]
//! extern crate serde_derive;
//!
//! use avro::Codec;
//! use avro::from_value;
//! use avro::reader::Reader;
//! use avro::schema::Schema;
//! use avro::types::Record;
//! use avro::writer::Writer;
//!
//! #[derive(Debug, Deserialize, Serialize)]
//! struct Test {
//! a: i64,
//! b: String,
//! }
//!
//! fn main() {
//! let raw_schema = r#"
//! {
//! "type": "record",
//! "name": "test",
//! "fields": [
//! {"name": "a", "type": "long", "default": 42},
//! {"name": "b", "type": "string"}
//! ]
//! }
//! "#;
//!
//! let schema = Schema::parse_str(raw_schema).unwrap();
//!
//! println!("{:?}", schema);
//!
//! let mut writer = Writer::with_codec(&schema, Vec::new(), Codec::Deflate);
//!
//! let mut record = Record::new(writer.schema()).unwrap();
//! record.put("a", 27i64);
//! record.put("b", "foo");
//!
//! writer.append(record).unwrap();
//!
//! let test = Test {
//! a: 27,
//! b: "foo".to_owned(),
//! };
//!
//! writer.append_ser(test).unwrap();
//!
//! let input = writer.into_inner();
//! let reader = Reader::with_schema(&schema, &input[..]);
//!
//! for record in reader {
//! println!("{:?}", from_value::<Test>(&record));
//! }
//! }
//! ```

extern crate failure;
extern crate libflate;
extern crate rand;
#[macro_use]
extern crate serde;
extern crate serde_json;
#[cfg(feature = "snappy")]
extern crate snap;

pub mod codec;
pub mod de;
pub mod decode;
pub mod encode;
pub mod reader;
pub mod schema;
pub mod ser;
pub mod types;
mod util;
pub mod writer;

pub use codec::Codec;

pub use de::from_value;

#[cfg(test)]
mod tests {

 #[test]
 fn it_works() {
 assert_eq!(2 + 2, 4);
 }
}
